src/armado/sqlite_index.py

```diff
@@ -157,3 +172,231 @@ def to_filename(title):
+            decomp_data = decompress_data(row[1])
+            for doc in decomp_data:
+                yield doc
+
+    @lru_cache(1)
+    def __len__(self):
+        """Compute the total number of docs in compressed pages."""
+        sql = "Select pageid, data from docs order by pageid desc limit 1"
+        cur = self.db.execute(sql)
+        row = cur.fetchone()
+        decomp_data = decompress_data(row[1])
+        return row[0] * PAGE_SIZE + len(decomp_data)
+
+    def random(self):
+        """Returns a random value."""
+        docid = random.randint(0, len(self) - 1)
+        return self.get_doc(docid)
+
+    def __contains__(self, key):
+        """Returns if the key is in the index or not."""
+        cur = self.db.execute("SELECT word FROM tokens where word = ?", (key,))
+        if cur.fetchone():
+            return True
+        return False
+
+    @lru_cache(1000)
+    def _get_page(self, pageid):
+        cur = self.db.execute("SELECT data FROM docs where pageid = ?", (pageid,))
+        row = cur.fetchone()
+        if row:
+            decomp_data = decompress_data(row[0])
+            return decomp_data
+        return None
+
+    def get_doc(self, docid):
+        '''Returns one stored document item.'''
+        page_id, rel_position = divmod(docid, PAGE_SIZE)
+        data = self._get_page(page_id)
+        if not data:
+            return None
+        row = data[rel_position]
+        # if the html filename is marked as computable
+        # do it and store in position 0.
+        if row[0] is None:
+            row[0] = to_filename(row[1])
+        return row
+
+    @classmethod
+    def create(cls, directory, source):
+        """Creates the index in the directory.
+        The source must give path, page_score, title and
+        a list of extracted words from title in an ordered fashion
+
+        It must return the quantity of pairs indexed.
+        """
+        import pickletools
+        import timeit
+
+        class SQLmany:
+            """Execute many INSERTs greatly improves the performance."""
+            def __init__(self, name, sql, quantity):
+                self.sql = sql
+                self.name = name
+                self.count = 0
+                self.buffer = []
+                self.progress_bar = ProgressBar(name, quantity)
+
+            def append(self, data):
+                """Append one data set to persist on db."""
+                self.buffer.append(data)
+                self.count += 1
+                if self.count % PAGE_SIZE == 0:
+                    self.persist()
+                    self.buffer = []
+                # self.count is the quantity of docs added
+                # but it is the index that is returned
+                # and it is zero based, hence one less.
+                self.progress_bar.step(self.count)
+                return self.count - 1
+
+            def finish(self):
+                """Finish the process and prints some data."""
```

Comment:
"prints some data"?

Reply:
logs some data... may be

Reply:
Pero qué info loguea o muestra? No está mostrando nada... o sí?